bundles/org.openhab.core.addon.marketplace/src/main/java/org/openhab/core/addon/marketplace/AbstractRemoteAddonService.java

```diff
@@ -0,0 +1,205 @@
+/**
+ * Copyright (c) 2010-2022 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.core.addon.marketplace;
+
+import java.io.IOException;
+import java.net.URI;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Dictionary;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.core.addon.Addon;
+import org.openhab.core.addon.AddonEventFactory;
+import org.openhab.core.addon.AddonService;
+import org.openhab.core.addon.AddonType;
+import org.openhab.core.cache.ExpiringCache;
+import org.openhab.core.events.Event;
+import org.openhab.core.events.EventPublisher;
+import org.openhab.core.storage.Storage;
+import org.openhab.core.storage.StorageService;
+import org.osgi.service.cm.Configuration;
+import org.osgi.service.cm.ConfigurationAdmin;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+/**
+ * The {@link AbstractRemoteAddonService} implements basic functionality of a remote add-on-service
+ *
+ * @author Jan N. Klug - Initial contribution
+ */
+@NonNullByDefault
+public abstract class AbstractRemoteAddonService implements AddonService {
+    protected static final Map<String, AddonType> TAG_ADDON_TYPE_MAP = Map.of( //
+            "automation", new AddonType("automation", "Automation"), //
+            "binding", new AddonType("binding", "Bindings"), //
+            "misc", new AddonType("misc", "Misc"), //
+            "persistence", new AddonType("persistence", "Persistence"), //
+            "transformation", new AddonType("transformation", "Transformations"), //
+            "ui", new AddonType("ui", "User Interfaces"), //
+            "voice", new AddonType("voice", "Voice"));
+
+    protected final Gson gson = new GsonBuilder().setDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'").create();
+    protected final Set<MarketplaceAddonHandler> addonHandlers = new HashSet<>();
+    protected final Storage<String> installedAddonStorage;
+    protected final EventPublisher eventPublisher;
+    protected final ConfigurationAdmin configurationAdmin;
+    protected final ExpiringCache<List<Addon>> cachedRemoteAddons = new ExpiringCache<>(Duration.ofMinutes(15),
+            this::getRemoteAddons);
+    protected List<Addon> cachedAddons = List.of();
```

Comment:
With the `ExpiringCache` in place this list looks a little bit duplicate. Isn't it?

If I suggest to change it to a `Map<String, Addon>` which allows an even faster access to an add-on in the `getAddon()` method if we choose the add-on ID as key for the `Map`.

Reply:
`cachedAddons` also contains the locally installed add-ons which are not available remotely, while `cachedRemoteAddons` only contains the remotely available add-ons. So it's not the same content.

Unfortunately `getAddons` requires a `List`, so it's not possible to use `cachedAddons.values()` and a `new ArrayList<>(cachedAddons.values())` is needed. So we trade the construction of a new `ArrayList` against the lookip in the `List`. Depending on the amount of calls to `getAddons` and `getAddon` this solution might be better or worse. I can change it if you want.

Reply:
If you don't mind the breaking change in API (which probably is not much of an issue), we could change `AddonService.getAddons` to return `Collection<Addon>`. At the moment there are only two calls to that method (in `org.openhab.core.io.rest.core.internal.addons.AddonResource` and both don't rely on the result being a `List` but call `.stream()` on the returned object.